File: 01-数据结构/01-数组/c++/LC 长度最小的子数组.cpp
#include<vector>

using namespace std;

class Solution {
public:
    int minSubArrayLen(int s, vector<int>& nums) {
        int l = 0;
        int r = 0;
        int min = nums.size() + 1;
        int sum = 0;
        while (r < nums.size() + 1) {
            if (sum < s) {
                if (r < nums.size()) {
                    sum += nums[r];
                }
                r++;
            }else {
                // cout<<l<<":"<<r<<"->"<<sum<<endl;
                if (r - l < min) {
                    min = r - l;
                }
                sum -= nums[l];
                l++;
            }
        }

        return min > nums.size()? 0 : min;

    }


    int minSubArrayLen2(int s, vector<int>& nums) {
        int l = 0;
        int r = 0;
        int min = nums.size() + 1;
        int sum = 0;
        while (l < nums.size()) {
            if (sum < s) {
                if (r == nums.size()) {
                    if (l == 0) {
                        return 0;
                    }else {
                        return min;
                    }
                }
                sum += nums[r];
                r++;
            }else {
                
                if (r - l < min) {
                    min = r - l;
                }
                // cout<<l<<":"<<r<<"->"<<sum<<" "<<min<<endl;

                sum -= nums[l];
                l++;
            }
        }

        return min > nums.size()? 0 : min;

    }
};

File: 01-数据结构/01-数组/go/LC 长度最小的子数组.go
package main

func minSubArrayLen(s int, nums []int) int {
    l, r := 0, 0
    sum := 0
    min := len(nums) + 1
    for l < len(nums) {
        
        sum = getsum(nums[l:r])
        // fmt.Println(nums[l:r],sum,sum == s, min)

        if sum >= s {
            if r == l {
                return 1;
            }
            if r - l < min {
                min = r - l
            }
            l++
        }else {
            if r < len(nums) {
                r++                
            }else {
                break;
            }
        }

    }

    if min > len(nums) {
        return 0
    }
    return min
}

func getsum(nums []int) (sum int) {
    for _, v := range nums {
        sum += v
    }

    return
}




func minSubArrayLen2(s int, nums []int) int {
    l, r := 0, 0
    sum := 0
    min := len(nums) + 1

    for r < len(nums) + 1 {
        // fmt.Println(nums[l:r],sum)

        if sum < s {
            if r < len(nums) {
                sum += nums[r]
            }
            r++
        }else {
            if r - l < min {
                min = r - l
            }
            sum -= nums[l]
            l++
        }
    }

    if min > len(nums) {
        return 0
    }else {
        return min
    }

}
